### PR TITLE
Javascript uploader

### PR DESCRIPTION
**Pull request by bcouston**
Also added progress bar images from resumable.js GitHub and
restructured visually. At the moment, the upload button is missing the
upload image, can be corrected later.

### CODE REVIEW

**Comment by bcouston** [comment]
DO NOT MERGE - Needs some review and maybe a few additional features.

**Comment by pezholio** [comment]
First impressions are pretty good - it seems to work pretty well, even with big files :+1: 

First thing I've noticed is the percentage feedback to users looks a bit odd - i.e.

mg3tlofolskqw6gilq4z0h_am-ejphaopn_jhblskma

I'd expect a percentage there really?

Also, the styling on the button looks a bit odd - there's a drop shadow there which wasn't there before

**Comment by pezholio** [comment]
Also, slight UI point as well - I'd do the upload after the user clicks validate - it's not 100% clear to me I need to do something after the file has uploaded.

**Comment by bcouston** [comment]
@pezholio, the thing on the right of the progress bar is a pause button for the upload, I'm not sure why it's encapsulated in an `<a>` tag...

Also by changing the file_field_tag for the upload button to a button tag which triggers some js to be able to browse for files, I had to change the styling a bit, that's why it looks a bit odd, but i will work on those parts :)

**Comment by pezholio** [comment]
Right - this is good to go I think! :+1:

File: app/views/application/index.html.erb [hunk 1]
```diff
@@ -1,6 +1,8 @@
 <% content_for :title do %>
   Check your CSV files with CSVLint
 <% end %>
+<%= javascript_include_tag "resumable.js" %>
+<%= stylesheet_link_tag "resumable.css" %>
```
Comment (Floppy) [inline app/views/application/index.html.erb]: We could include these into the asset pipeline with in application.js and application.css

File: app/views/application/index.html.erb [hunk 2]
```diff
@@ -53,6 +74,10 @@
 
       <p class="more-information">Submitted urls are recorded in a public <a href="/validation">list of validation reports</a>. If you want to validate private data then upload a file from your computer, using the Browse button below.</p>
       <% end %>
+    <div class="resumable-error">
+      Your browser, unfortunately, is not supported by Resumable.js. The library requires support for <a href="http://www.w3.org/TR/FileAPI/">the HTML5 File API</a> along with <a href="http://www.w3.org/TR/FileAPI/#normalization-of-params">file slicing</a>.
+    </div>
+
```
Comment (Floppy) [inline app/views/application/index.html.erb]: What happens if it's not supported? Do we fall back to standard file upload? Try testing in a browser with javascript disabled to find out.

File: lib/stored_csv.rb [hunk 3]
```diff
@@ -0,0 +1,15 @@
+require 'mongoid/grid_fs'
+
+class StoredCSV
```
Comment (Floppy) [inline lib/stored_csv.rb]: could this wrap the retrieval as well as the storage?

Comment (pezholio) [inline lib/stored_csv.rb]: Totally. I meant to do this anyway

**Comment by Floppy** [comment]
OK, this looks all sensible, but I have one concern. What happens if the resumable.js requests to go different nodes? Or if one dies partway through? It seems like chunks are stored on the filesystem during upload, then moved to GridFs on completion, but that means that it might not work on multiple nodes.

**Comment by pezholio** [comment]
Oh, good shout. I didn't think about that. In that case, we could do it all in GridFS, then throw away the chunks.

**Comment by Floppy** [comment]
Just discussed with @bcouston, and he's going to update the engine to handle that case, then we'll merge this PR once it uses the engine.

**Comment by pezholio** [comment]
Cool :+1: I have also noticed that it doesn't work at all with JS disabled now. We might need to look at that too

**Comment by Floppy** [comment]
OK, fab. Can you do me a favour and make tickets for those two tasks in huboard so we can include them in sprint planning?

**Pull request merged**